# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [2]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [3]:
view_sentence_range = (32, 42)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 32 to 42:
your most feared animal is that shark .
paris is usually wet during august , and it is never dry in november .
paris is usually beautiful during september , and it is usually snowy in november .
the united states is never wet during january , but it is usually hot in october .
we like oranges , mangoes , and grapes .
they like pears , apples , and mangoes .
she dislikes that little red truck .
the grapefruit is my most loved fruit , but the banana is her most loved .
france is snowy during may , and it is never busy in autumn .
china is usually mild during winter , but it is never busy in february .

French sentences 32 to 42:
votre animal le plus redouté est que le requin .
paris est généralement humide au mois d' août , et il est jamais sec en novembre .
paris est généralement beau en septembre , et il est généralemen

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [4]:

def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    source_lines = [line for line in source_text.split('\n')]
    target_lines = [line + ' <EOS>' for line in target_text.split('\n')]
    
    source_id_text = [[source_vocab_to_int[word] for word in line.split()] for line in source_lines]
    target_id_text = [[target_vocab_to_int[word] for word in line.split()] for line in target_lines]
    
    return source_id_text, target_id_text

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [7]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.1.0


C:\Users\wzhao\AppData\Local\Continuum\Anaconda3\envs\py3\lib\site-packages\ipykernel\__main__.py:15: UserWarning: No GPU found. Please use a GPU to train your neural network.


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [8]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    # TODO: Implement Function
    input = tf.placeholder(tf.int32, [None, None], name='input')
    targets = tf.placeholder(tf.int32, [None, None], name='targets')
    learning_rate = tf.placeholder(tf.float32, name='lr')
    target_sequence_length = tf.placeholder(tf.int32, [None], name='target_sequence_length')
    source_sequence_len = tf.placeholder(tf.int32, [None], name='source_sequence_length')
    max_target_len = tf.reduce_max(target_sequence_length, name= "max_target_len")    
    keep_prob = tf.placeholder(tf.float32, name='keep_prob')
    return input, targets, learning_rate, keep_prob, target_sequence_length, max_target_len, source_sequence_len


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [9]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    # TODO: Implement Function
    targets = tf.strided_slice(target_data, [0, 0], [batch_size, -1], [1, 1])
    return tf.concat([tf.fill([batch_size, 1], target_vocab_to_int['<GO>']), targets], 1)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [10]:
from imp import reload
reload(tests)

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    # TODO: Implement Function
    enc_embed = tf.contrib.layers.embed_sequence(rnn_inputs, source_vocab_size, encoding_embedding_size)
    
    def create_cell(rnn_size):
        lstm_cell = tf.contrib.rnn.LSTMCell(rnn_size,initializer=tf.random_uniform_initializer(-0.1,0.1,seed=2))
        drop = tf.contrib.rnn.DropoutWrapper(lstm_cell, output_keep_prob=keep_prob)
        return drop
    
    enc_cell = tf.contrib.rnn.MultiRNNCell([create_cell(rnn_size) for _ in range(num_layers)])
    encoding_output, encoding_state = tf.nn.dynamic_rnn(enc_cell, enc_embed, 
                                                        sequence_length=source_sequence_length,dtype=tf.float32)
    
    return encoding_output, encoding_state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [11]:

def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    # TODO: Implement Function
    helper = tf.contrib.seq2seq.TrainingHelper(inputs=dec_embed_input, sequence_length=target_sequence_length)
    training_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell, helper, encoder_state, output_layer)
    decoded, _ = tf.contrib.seq2seq.dynamic_decode(training_decoder, 
                                                      impute_finished=True, maximum_iterations=max_summary_length)
    
    # Return the decoded outputs
    return decoded



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [12]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    # TODO: Implement Function
    tokens = tf.tile(tf.constant([start_of_sequence_id], dtype=tf.int32), [batch_size], name='start_tokens')
    helper = tf.contrib.seq2seq.GreedyEmbeddingHelper(dec_embeddings, tokens, end_of_sequence_id)
    decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell, helper, encoder_state, output_layer)
    decoder_output, _ = tf.contrib.seq2seq.dynamic_decode(decoder, impute_finished=True, 
                                                         maximum_iterations=max_target_sequence_length)       

    return decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [15]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :param decoding_embedding_size: Decoding embedding size
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function

    dec_embeddings = tf.Variable(tf.random_uniform([target_vocab_size, decoding_embedding_size]))
    dec_embed_input = tf.nn.embedding_lookup(dec_embeddings, dec_input)
    
 
    def create_decoder_cell(rnn_size):
        lstm_cell = tf.contrib.rnn.LSTMCell(rnn_size,
                                            initializer=tf.random_uniform_initializer(-0.1,0.1,seed=2))
        drop = tf.contrib.rnn.DropoutWrapper(lstm_cell, output_keep_prob=keep_prob)
        return drop

    dec_cell = tf.contrib.rnn.MultiRNNCell([create_decoder_cell(rnn_size) for _ in range(num_layers)])
    
    output_layer = Dense(target_vocab_size,
                         kernel_initializer = tf.truncated_normal_initializer(mean = 0.0, stddev=0.1))

    with tf.variable_scope("decode"):
        train_decoder_out = decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_target_sequence_length, output_layer, keep_prob)
        
    with tf.variable_scope("decode", reuse=True):
        infer_decoder_out = decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, 
                             target_vocab_to_int['<GO>'], target_vocab_to_int['<EOS>'], max_target_sequence_length, 
                             target_vocab_size, output_layer, batch_size, keep_prob)
        
    return (train_decoder_out, infer_decoder_out)



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [16]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    encoded, enc_state = encoding_layer(input_data, rnn_size, num_layers, keep_prob, source_sequence_length, source_vocab_size, 
                                        enc_embedding_size)
    dec_in = process_decoder_input(target_data, target_vocab_to_int, batch_size)

    train, infer = decoding_layer(dec_in, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, 
                                  num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)
    
    # Return the results
    return train, infer


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [17]:
# Number of Epochs
epochs = 10
# Batch Size
batch_size = 156
# RNN Size
rnn_size = 256
# Number of Layers
num_layers = 2
# Embedding Size
encoding_embedding_size = 128
decoding_embedding_size = 128
# Learning Rate
learning_rate = 0.001
# Dropout Keep Probability
keep_probability = 0.75
display_step = 10

### Build the Graph
Build the graph using the neural network you implemented.

In [18]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [19]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [20]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch   10/883 - Train Accuracy: 0.3234, Validation Accuracy: 0.3802, Loss: 3.5014
Epoch   0 Batch   20/883 - Train Accuracy: 0.3647, Validation Accuracy: 0.4178, Loss: 3.0564
Epoch   0 Batch   30/883 - Train Accuracy: 0.4000, Validation Accuracy: 0.4525, Loss: 2.7343
Epoch   0 Batch   40/883 - Train Accuracy: 0.4032, Validation Accuracy: 0.4481, Loss: 2.5543
Epoch   0 Batch   50/883 - Train Accuracy: 0.4163, Validation Accuracy: 0.4677, Loss: 2.3858
Epoch   0 Batch   60/883 - Train Accuracy: 0.4337, Validation Accuracy: 0.4933, Loss: 2.2333
Epoch   0 Batch   70/883 - Train Accuracy: 0.4433, Validation Accuracy: 0.4921, Loss: 2.1102
Epoch   0 Batch   80/883 - Train Accuracy: 0.4936, Validation Accuracy: 0.5198, Loss: 1.9335
Epoch   0 Batch   90/883 - Train Accuracy: 0.4638, Validation Accuracy: 0.5003, Loss: 1.7922
Epoch   0 Batch  100/883 - Train Accuracy: 0.4506, Validation Accuracy: 0.4939, Loss: 1.6948
Epoch   0 Batch  110/883 - Train Accuracy: 0.4701, Validation Accuracy

Epoch   1 Batch   20/883 - Train Accuracy: 0.8324, Validation Accuracy: 0.8124, Loss: 0.2769
Epoch   1 Batch   30/883 - Train Accuracy: 0.8567, Validation Accuracy: 0.8260, Loss: 0.2624
Epoch   1 Batch   40/883 - Train Accuracy: 0.8165, Validation Accuracy: 0.8395, Loss: 0.2614
Epoch   1 Batch   50/883 - Train Accuracy: 0.8167, Validation Accuracy: 0.8022, Loss: 0.2672
Epoch   1 Batch   60/883 - Train Accuracy: 0.8048, Validation Accuracy: 0.8295, Loss: 0.2268
Epoch   1 Batch   70/883 - Train Accuracy: 0.8330, Validation Accuracy: 0.8400, Loss: 0.2359
Epoch   1 Batch   80/883 - Train Accuracy: 0.8397, Validation Accuracy: 0.8362, Loss: 0.2354
Epoch   1 Batch   90/883 - Train Accuracy: 0.8327, Validation Accuracy: 0.8211, Loss: 0.2087
Epoch   1 Batch  100/883 - Train Accuracy: 0.8433, Validation Accuracy: 0.8258, Loss: 0.2115
Epoch   1 Batch  110/883 - Train Accuracy: 0.8614, Validation Accuracy: 0.8234, Loss: 0.1971
Epoch   1 Batch  120/883 - Train Accuracy: 0.8480, Validation Accuracy

Epoch   2 Batch   30/883 - Train Accuracy: 0.9413, Validation Accuracy: 0.9324, Loss: 0.0660
Epoch   2 Batch   40/883 - Train Accuracy: 0.9396, Validation Accuracy: 0.9225, Loss: 0.0679
Epoch   2 Batch   50/883 - Train Accuracy: 0.9362, Validation Accuracy: 0.9143, Loss: 0.0816
Epoch   2 Batch   60/883 - Train Accuracy: 0.9234, Validation Accuracy: 0.9199, Loss: 0.0679
Epoch   2 Batch   70/883 - Train Accuracy: 0.9205, Validation Accuracy: 0.9274, Loss: 0.0680
Epoch   2 Batch   80/883 - Train Accuracy: 0.9277, Validation Accuracy: 0.9274, Loss: 0.0660
Epoch   2 Batch   90/883 - Train Accuracy: 0.9298, Validation Accuracy: 0.9225, Loss: 0.0549
Epoch   2 Batch  100/883 - Train Accuracy: 0.9494, Validation Accuracy: 0.9164, Loss: 0.0554
Epoch   2 Batch  110/883 - Train Accuracy: 0.9325, Validation Accuracy: 0.9111, Loss: 0.0584
Epoch   2 Batch  120/883 - Train Accuracy: 0.9145, Validation Accuracy: 0.9263, Loss: 0.0870
Epoch   2 Batch  130/883 - Train Accuracy: 0.9457, Validation Accuracy

Epoch   3 Batch   40/883 - Train Accuracy: 0.9560, Validation Accuracy: 0.9365, Loss: 0.0448
Epoch   3 Batch   50/883 - Train Accuracy: 0.9439, Validation Accuracy: 0.9350, Loss: 0.0495
Epoch   3 Batch   60/883 - Train Accuracy: 0.9462, Validation Accuracy: 0.9353, Loss: 0.0396
Epoch   3 Batch   70/883 - Train Accuracy: 0.9478, Validation Accuracy: 0.9350, Loss: 0.0391
Epoch   3 Batch   80/883 - Train Accuracy: 0.9499, Validation Accuracy: 0.9309, Loss: 0.0436
Epoch   3 Batch   90/883 - Train Accuracy: 0.9500, Validation Accuracy: 0.9324, Loss: 0.0334
Epoch   3 Batch  100/883 - Train Accuracy: 0.9708, Validation Accuracy: 0.9222, Loss: 0.0373
Epoch   3 Batch  110/883 - Train Accuracy: 0.9719, Validation Accuracy: 0.9181, Loss: 0.0300
Epoch   3 Batch  120/883 - Train Accuracy: 0.9432, Validation Accuracy: 0.9385, Loss: 0.0580
Epoch   3 Batch  130/883 - Train Accuracy: 0.9625, Validation Accuracy: 0.9411, Loss: 0.0348
Epoch   3 Batch  140/883 - Train Accuracy: 0.9522, Validation Accuracy

Epoch   4 Batch   50/883 - Train Accuracy: 0.9644, Validation Accuracy: 0.9496, Loss: 0.0294
Epoch   4 Batch   60/883 - Train Accuracy: 0.9561, Validation Accuracy: 0.9467, Loss: 0.0297
Epoch   4 Batch   70/883 - Train Accuracy: 0.9590, Validation Accuracy: 0.9467, Loss: 0.0286
Epoch   4 Batch   80/883 - Train Accuracy: 0.9493, Validation Accuracy: 0.9417, Loss: 0.0287
Epoch   4 Batch   90/883 - Train Accuracy: 0.9821, Validation Accuracy: 0.9569, Loss: 0.0254
Epoch   4 Batch  100/883 - Train Accuracy: 0.9731, Validation Accuracy: 0.9677, Loss: 0.0299
Epoch   4 Batch  110/883 - Train Accuracy: 0.9643, Validation Accuracy: 0.9604, Loss: 0.0220
Epoch   4 Batch  120/883 - Train Accuracy: 0.9521, Validation Accuracy: 0.9493, Loss: 0.0440
Epoch   4 Batch  130/883 - Train Accuracy: 0.9576, Validation Accuracy: 0.9449, Loss: 0.0300
Epoch   4 Batch  140/883 - Train Accuracy: 0.9799, Validation Accuracy: 0.9516, Loss: 0.0259
Epoch   4 Batch  150/883 - Train Accuracy: 0.9615, Validation Accuracy

Epoch   5 Batch   60/883 - Train Accuracy: 0.9657, Validation Accuracy: 0.9636, Loss: 0.0215
Epoch   5 Batch   70/883 - Train Accuracy: 0.9654, Validation Accuracy: 0.9639, Loss: 0.0270
Epoch   5 Batch   80/883 - Train Accuracy: 0.9505, Validation Accuracy: 0.9580, Loss: 0.0256
Epoch   5 Batch   90/883 - Train Accuracy: 0.9804, Validation Accuracy: 0.9578, Loss: 0.0152
Epoch   5 Batch  100/883 - Train Accuracy: 0.9798, Validation Accuracy: 0.9537, Loss: 0.0216
Epoch   5 Batch  110/883 - Train Accuracy: 0.9838, Validation Accuracy: 0.9557, Loss: 0.0161
Epoch   5 Batch  120/883 - Train Accuracy: 0.9652, Validation Accuracy: 0.9522, Loss: 0.0406
Epoch   5 Batch  130/883 - Train Accuracy: 0.9689, Validation Accuracy: 0.9583, Loss: 0.0230
Epoch   5 Batch  140/883 - Train Accuracy: 0.9787, Validation Accuracy: 0.9636, Loss: 0.0222
Epoch   5 Batch  150/883 - Train Accuracy: 0.9679, Validation Accuracy: 0.9677, Loss: 0.0230
Epoch   5 Batch  160/883 - Train Accuracy: 0.9708, Validation Accuracy

Epoch   6 Batch   70/883 - Train Accuracy: 0.9622, Validation Accuracy: 0.9566, Loss: 0.0271
Epoch   6 Batch   80/883 - Train Accuracy: 0.9637, Validation Accuracy: 0.9595, Loss: 0.0246
Epoch   6 Batch   90/883 - Train Accuracy: 0.9907, Validation Accuracy: 0.9636, Loss: 0.0119
Epoch   6 Batch  100/883 - Train Accuracy: 0.9808, Validation Accuracy: 0.9752, Loss: 0.0242
Epoch   6 Batch  110/883 - Train Accuracy: 0.9918, Validation Accuracy: 0.9490, Loss: 0.0169
Epoch   6 Batch  120/883 - Train Accuracy: 0.9695, Validation Accuracy: 0.9674, Loss: 0.0377
Epoch   6 Batch  130/883 - Train Accuracy: 0.9741, Validation Accuracy: 0.9744, Loss: 0.0179
Epoch   6 Batch  140/883 - Train Accuracy: 0.9819, Validation Accuracy: 0.9580, Loss: 0.0185
Epoch   6 Batch  150/883 - Train Accuracy: 0.9824, Validation Accuracy: 0.9653, Loss: 0.0194
Epoch   6 Batch  160/883 - Train Accuracy: 0.9635, Validation Accuracy: 0.9621, Loss: 0.0211
Epoch   6 Batch  170/883 - Train Accuracy: 0.9869, Validation Accuracy

Epoch   7 Batch   80/883 - Train Accuracy: 0.9579, Validation Accuracy: 0.9656, Loss: 0.0214
Epoch   7 Batch   90/883 - Train Accuracy: 0.9744, Validation Accuracy: 0.9700, Loss: 0.0110
Epoch   7 Batch  100/883 - Train Accuracy: 0.9811, Validation Accuracy: 0.9802, Loss: 0.0158
Epoch   7 Batch  110/883 - Train Accuracy: 0.9820, Validation Accuracy: 0.9694, Loss: 0.0156
Epoch   7 Batch  120/883 - Train Accuracy: 0.9698, Validation Accuracy: 0.9738, Loss: 0.0322
Epoch   7 Batch  130/883 - Train Accuracy: 0.9737, Validation Accuracy: 0.9761, Loss: 0.0165
Epoch   7 Batch  140/883 - Train Accuracy: 0.9796, Validation Accuracy: 0.9566, Loss: 0.0147
Epoch   7 Batch  150/883 - Train Accuracy: 0.9939, Validation Accuracy: 0.9767, Loss: 0.0141
Epoch   7 Batch  160/883 - Train Accuracy: 0.9612, Validation Accuracy: 0.9761, Loss: 0.0187
Epoch   7 Batch  170/883 - Train Accuracy: 0.9857, Validation Accuracy: 0.9758, Loss: 0.0167
Epoch   7 Batch  180/883 - Train Accuracy: 0.9956, Validation Accuracy

Epoch   8 Batch   90/883 - Train Accuracy: 0.9894, Validation Accuracy: 0.9662, Loss: 0.0097
Epoch   8 Batch  100/883 - Train Accuracy: 0.9913, Validation Accuracy: 0.9793, Loss: 0.0177
Epoch   8 Batch  110/883 - Train Accuracy: 0.9722, Validation Accuracy: 0.9557, Loss: 0.0181
Epoch   8 Batch  120/883 - Train Accuracy: 0.9728, Validation Accuracy: 0.9604, Loss: 0.0337
Epoch   8 Batch  130/883 - Train Accuracy: 0.9768, Validation Accuracy: 0.9741, Loss: 0.0177
Epoch   8 Batch  140/883 - Train Accuracy: 0.9825, Validation Accuracy: 0.9639, Loss: 0.0191
Epoch   8 Batch  150/883 - Train Accuracy: 0.9910, Validation Accuracy: 0.9720, Loss: 0.0202
Epoch   8 Batch  160/883 - Train Accuracy: 0.9769, Validation Accuracy: 0.9747, Loss: 0.0165
Epoch   8 Batch  170/883 - Train Accuracy: 0.9802, Validation Accuracy: 0.9755, Loss: 0.0148
Epoch   8 Batch  180/883 - Train Accuracy: 0.9865, Validation Accuracy: 0.9548, Loss: 0.0123
Epoch   8 Batch  190/883 - Train Accuracy: 0.9936, Validation Accuracy

Epoch   9 Batch  100/883 - Train Accuracy: 0.9897, Validation Accuracy: 0.9723, Loss: 0.0129
Epoch   9 Batch  110/883 - Train Accuracy: 0.9973, Validation Accuracy: 0.9642, Loss: 0.0089
Epoch   9 Batch  120/883 - Train Accuracy: 0.9771, Validation Accuracy: 0.9720, Loss: 0.0248
Epoch   9 Batch  130/883 - Train Accuracy: 0.9884, Validation Accuracy: 0.9662, Loss: 0.0123
Epoch   9 Batch  140/883 - Train Accuracy: 0.9918, Validation Accuracy: 0.9781, Loss: 0.0116
Epoch   9 Batch  150/883 - Train Accuracy: 0.9824, Validation Accuracy: 0.9624, Loss: 0.0145
Epoch   9 Batch  160/883 - Train Accuracy: 0.9766, Validation Accuracy: 0.9822, Loss: 0.0143
Epoch   9 Batch  170/883 - Train Accuracy: 0.9841, Validation Accuracy: 0.9802, Loss: 0.0140
Epoch   9 Batch  180/883 - Train Accuracy: 0.9872, Validation Accuracy: 0.9691, Loss: 0.0120
Epoch   9 Batch  190/883 - Train Accuracy: 0.9865, Validation Accuracy: 0.9799, Loss: 0.0101
Epoch   9 Batch  200/883 - Train Accuracy: 0.9898, Validation Accuracy

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [21]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [22]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [23]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    # TODO: Implement Function
    unk = vocab_to_int['<UNK>']
    return [vocab_to_int.get(word, unk) for word in sentence.lower().split()]


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [24]:
translate_sentence = 'he saw a old yellow truck .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [4, 122, 88, 25, 172, 212, 216]
  English Words: ['he', 'saw', 'a', 'old', 'yellow', 'truck', '.']

Prediction
  Word Ids:      [237, 299, 275, 159, 289, 153, 38, 352, 1]
  French Words: il a vu un vieux camion jaune . <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.